File: 56.merge-intervals/solution.kt
class Solution {
    fun merge(intervals: Array<IntArray>): Array<IntArray> {
        val seq = mutableListOf<IntArray>()
        var lastInterval = intArrayOf(0, 0)
        for (interval in intervals.sortedBy { it[0] }) {
            if (seq.isEmpty() || interval[0] > lastInterval[1]) {
                seq += interval
                lastInterval = interval
            } else {
                lastInterval[1] = maxOf(lastInterval[1], interval[1])
            }
        } 
        return seq.toTypedArray()
    }
}

File: 56.merge-intervals/solution.scala
object Solution {
    def merge(intervals: Array[Array[Int]]): Array[Array[Int]] = {
        intervals
            .sortBy(_(0))
            .foldLeft(List.empty[Array[Int]])((seq, interval) =>
                if (seq.isEmpty || interval(0) > seq.last(1)) {
                    seq :+ interval
                } else {
                    seq.last(1) = Math.max(seq.last(1), interval(1))
                    seq
                }
            )
            .toArray
    }
}